internal/command/diff_inputs.go

```diff
@@ -0,0 +1,387 @@
+
+	"github.com/simplesurance/baur/v1"
+	"github.com/simplesurance/baur/v1/internal/digest"
+	"github.com/simplesurance/baur/v1/internal/format"
+	"github.com/simplesurance/baur/v1/internal/format/csv"
+	"github.com/simplesurance/baur/v1/internal/format/table"
+	"github.com/simplesurance/baur/v1/storage"
+)
+
+type storageInput struct {
+	input *storage.Input
+}
+
+func (i *storageInput) Digest() (*digest.Digest, error) {
+	return digest.FromString(i.input.Digest)
+}
+
+func (i *storageInput) String() string {
+	return i.input.URI
+}
+
+type diffInputArgDetails struct {
+	arg      string
+	appName  string
+	taskName string
+	runID    string
+	task     *baur.Task
+}
+
+func init() {
+	diffCmd.AddCommand(&newDiffInputsCmd().Command)
+}
+
+type diffInputsCmd struct {
+	cobra.Command
+
+	csv      bool
+	quiet    bool
+	inputStr string
+}
+
+func newDiffInputsCmd() *diffInputsCmd {
+	cmd := diffInputsCmd{
+		Command: cobra.Command{
+			Use:   "inputs <APP-NAME>.<TASK-NAME>|<RUN-ID> <APP-NAME>.<TASK-NAME>|<RUN-ID>",
+			Short: "list inputs that differ between two task-runs",
+			Long: `if the inputs match exit code 0 is returned, exit code 2 if they are different or exit code 1 if an error occurs.
+when outputting the differences, State 'D' indicates the digests do not match, '+' indicates the input is missing from the first argument and '-' indicates the input is missing from the second argument`,
+			Args: diffArgs(),
+		},
+	}
+
+	cmd.Run = cmd.run
+
+	cmd.Flags().BoolVar(&cmd.csv, "csv", false,
+		"show output in RFC4180 CSV format")
+
+	cmd.Flags().BoolVarP(&cmd.quiet, "quiet", "q", false,
+		"do not show anything, if the inputs match exit code 0 is returned, exit code 2 if they are different or exit code 1 if an error occurs")
+
+	cmd.Flags().StringVar(&cmd.inputStr, "input-str", "",
+		"include a string as input")
+
+	return &cmd
+}
+
+// diffArgs returns an error in the following scenarios:
+// - there is less than or greater than 2 args specified
+// - the <APP-NAME> or <TASK-NAME> is a wildcard character (*)
+// - either arg is not in the format <APP-NAME>.<TASK-NAME> or a numeric value
+func diffArgs() cobra.PositionalArgs {
+	return func(cmd *cobra.Command, args []string) error {
+		if len(args) != 2 {
+			return fmt.Errorf("accepts 2 args, received %d", len(args))
+		}
+
+		containsWildCardPattern := "^\\*\\..+$|^.+\\.\\*$"
+		isNumericPattern := "^\\d+$"
+		containsTaskPattern := "^.+\\..+$"
+
+		for _, arg := range args {
+			containsWildCard, err := regexp.MatchString(containsWildCardPattern, arg)
+
+			if err != nil {
+				return err
+			}
+
+			if containsWildCard {
+				return fmt.Errorf("%s contains a wild card character, wild card characters are not allowed", arg)
+			}
+
+			isValid, err := regexp.MatchString(fmt.Sprintf("%s|%s", isNumericPattern, containsTaskPattern), arg)
+
+			if err != nil {
+				return err
+			}
+
+			if !isValid {
+				return fmt.Errorf("%s does not specify a task or task-run ID", arg)
+			}
+		}
```

Comment:
If you are happy with the reduced error output and the fact that `app_one.build*` results in a match and will not cause an error but `app_one.*build` does not result in a match and returns an error then we can use your implementation.
The current implementation matches the asterisk only on either side of the period.

Reply:
> If you are happy with the reduced error output

I think the reduced output is fine.

>  and the fact that app_one.build* results in a match and will not cause an error but app_one.*build does not result in a match and returns an error then we can use your implementation.

good catch, an end of string marker is needed after the  `<APP>.<TASK>` regex path and a beginning of string marker before the run-id part:
`^(\w+\.\w+)$|^[0-9]+\d*$`
https://regex101.com/r/D1DRNC/3



>  The current implementation matches the asterisk only on either side of the period.

What is reason for allowing to have an asterik?
With your implementation a call like `diff inputs -q random.build random.buil*` is allowed as valid argument.
But this can not resolve to the existing `random.build` task because this kind of syntax is not supported in the loader as wildcard.

Reply:
> With your implementation a call like `diff inputs -q random.build random.buil*` is allowed as valid argument.
> But this can not resolve to the existing `random.build` task because this kind of syntax is not supported in the loader as wildcard.

I was not aware that the app and task names cannot contain the asterisk character at all, they would have been treated literally instead of as a wildcard. I had taken from [here](https://github.com/simplesurance/baur/blob/master/internal/command/ls_runs.go#L192) that an asterisk by itself was not valid.
I'm more than happy for the asterisk character to be invalid in all cases, in which case your latest regex pattern looks good.

Reply:
`ls apps` currently supports a syntax like:
- `*.build` -> to get the status of all build tasks for all apps
- `calc.*` -> to get the status of all tasks of the calc app

It does not work to use `*` as a wildcard for parts of the app or task names, like `calc.b*` will match `calc.build` and `calc.bla`.

For the `diff inputs` command one parameter should always resolve to a single task.
Supporting the wildcard syntax same way then for "ls apps" does not add much value here.
We would then also need to handle the case that the loader returned multiple tasks for a single parameter and print an error.

Reply:
I'll simplify things now and make the change.
I have tweaked your latest regex a little so it will only match on the whole argument and allow carets after the task
https://regex101.com/r/2VdBoe/1